### DISCUSSION TITLE
It is difficult to use plugins that require app.use to load

### DISCUSSION THREAD

**Discussion by CrystalWindSnake** [zauberzeug/nicegui#1742] [Q&A]
### Question

I want to introduce vue router, but currently nicegui doesn't seem to have a suitable mechanism to correctly load plugins.

```python
from nicegui import ui

ui.add_body_html('<script src="https://unpkg.com/vue-router@4"></script>')

ui.add_body_html(
    r"""
<script >
           
document.addEventListener('DOMContentLoaded', () => {

    const Home = { template: '<div>Home</div>' }
    const About = { template: '<div>About</div>' }


    const routes = [
        { path: '/', component: Home },
        { path: '/about', component: About },
    ]


    const router = VueRouter.createRouter({
        history: VueRouter.createWebHashHistory(),
        routes, 
    })

    const appDom = document.getElementById('app')
    const app = appDom.__vue_app__
    app.use(router)
});
</script>
"""
)


@ui.refreshable
def create_router_view():
    ui.element("router-link").props('to="/" innerHTML="Go to Home"')
    ui.element("router-link").props('to="/about" innerHTML="Go to About"')

    # not working and I don't understand why .
    ui.element("router-view")


# not work,because the route configuration has not been executed yet
create_router_view()


def onclick():
    create_router_view.refresh()


ui.button("update router", on_click=onclick)

ui.run()
```

Is there any other more suitable way? Please tell me what I missed. Thank you.

**Comment by falkoschindler** [comment]
Yes, support for third-party libraries is still limited, see https://github.com/zauberzeug/nicegui/discussions/1454.
But maybe our [single page app example](https://github.com/zauberzeug/nicegui/tree/main/examples/single_page_app) is a good starting point for what you're trying to do?

**Reply by CrystalWindSnake** [reply]
I have already seen the official spa example and it is indeed a good implementation idea.
The reason why I tried to introduce vue router is that it contains many features, such as transition animations, navigation guards.If we can directly introduce vue router, control the spa's route navigation at the front end, and send the navigation information back to the backend (similar to event notification), and then execute the corresponding interface refresh function, it may be a way to implement a low-cost full spa routing.

Anyway, the extensibility of nicegui has been excellent so far.Thanks for all your work with nicegui... 👍
